linkerd/proxy/http/src/detect.rs

```diff
@@ -85,45 +41,108 @@ impl detect::Detect<tls::accept::Meta, BoxedIo> for DetectHttp {
 /// Otherwise, the `F` type forwarding service is used to handle the TCP
 /// connection.
 #[derive(Clone, Debug)]
-pub struct ServeHttp<F, H> {
-    http: hyper::server::conn::Http<trace::Executor>,
-    forward_tcp: F,
-    make_http: H,
+pub struct AcceptHttp<F, H> {
+    tcp: F,
+    http: H,
+    timeout: Duration,
+    server: hyper::server::conn::Http<trace::Executor>,
     drain: drain::Watch,
 }
 
-impl<F, H> ServeHttp<F, H> {
-    /// Creates a new `ServeHttp`.
-    pub fn new(make_http: H, h2: H2Settings, forward_tcp: F, drain: drain::Watch) -> Self {
-        let mut http = hyper::server::conn::Http::new().with_executor(trace::Executor::new());
+// === impl DetectHttp ===
 
-        http.http2_initial_stream_window_size(h2.initial_stream_window_size)
+impl<F, H> DetectHttp<F, H> {
+    /// Creates a new `AcceptHttp`.
+    pub fn new(h2: H2Settings, timeout: Duration, http: H, tcp: F, drain: drain::Watch) -> Self {
+        let mut server = hyper::server::conn::Http::new().with_executor(trace::Executor::new());
+        server
+            .http2_initial_stream_window_size(h2.initial_stream_window_size)
             .http2_initial_connection_window_size(h2.initial_connection_window_size);
 
         Self {
+            timeout,
+            server,
+            tcp,
             http,
-            forward_tcp,
-            make_http,
             drain,
         }
     }
 }
 
-impl<T, I, F, H, S> Service<(Protocol<T>, I)> for ServeHttp<F, H>
+impl<T, F, S> Service<T> for DetectHttp<F, S>
+where
+    T: Clone + Send + 'static,
+    F: tower::Service<T> + Clone + Send + 'static,
+    F::Error: Into<Error>,
+    F::Response: Send + 'static,
+    F::Future: Send + 'static,
+    S: Service<T> + Clone + Unpin + Send + 'static,
+    S::Error: Into<Error>,
+    S::Response: Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = AcceptHttp<F::Response, S::Response>;
+    type Error = Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let drain = self.drain.clone();
+        let tcp = self.tcp.clone();
+        let http = self.http.clone();
+        let server = self.server.clone();
+        let timeout = self.timeout;
+
+        Box::pin(async move {
+            let (tcp, http) = futures::try_join!(
+                tcp.oneshot(target.clone()).map_err(Into::<Error>::into),
+                http.oneshot(target).map_err(Into::<Error>::into)
```

Comment:
nit/TIOLI:
```suggestion
                tcp.oneshot(target.clone()).err_into::<Error>(),
                http.oneshot(target).err_into::<Error>(),
```
maybe slightly more concise. not a blocker though.

Reply:
iirc that didn't work